lib/teiserver/o_auth.ex

```diff
@@ -0,0 +1,419 @@
+    # the code for a token
+    attrs = %{
+      value: Base.hex_encode32(:crypto.strong_rand_bytes(32)),
+      owner_id: user_id,
+      application_id: attrs.id,
+      scopes: attrs.scopes,
+      expires_at: Timex.add(now, Timex.Duration.from_minutes(5)),
+      redirect_uri: Map.get(attrs, :redirect_uri),
+      challenge: Map.get(attrs, :challenge),
+      challenge_method: Map.get(attrs, :challenge_method)
+    }
+
+    %Code{}
+    |> Code.changeset(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Given a code returns the corresponding db object, making sure
+  it is valid (exists, not expired, not revoked)
+  """
+  @spec get_valid_code(String.t(), options()) :: {:ok, Code.t()} | {:error, term()}
+  def get_valid_code(code, opts \\ [])
+
+  def get_valid_code(code, opts) do
+    case CodeQueries.get_code(code) do
+      nil ->
+        {:error, :no_code}
+
+      code ->
+        now = Keyword.get(opts, :now, Timex.now())
+        check_expiry(code, now)
+    end
+  end
+
+  @spec create_token(
+          User.t() | T.userid(),
+          %{id: integer(), scopes: Application.scopes()},
+          create_refresh: boolean() | options()
+        ) ::
+          {:ok, Token.t()} | {:error, Ecto.Changeset.t()}
+  def create_token(user_id, application, opts \\ [])
+
+  def create_token(%User{} = user, application, opts) do
+    create_token(user.id, application, opts)
+  end
+
+  def create_token(user, application, opts) when is_map(user) do
+    create_token(user.id, application, opts)
+  end
+
+  def create_token(user_id, application, opts) do
+    do_create_token(%{owner_id: user_id}, application, opts)
+  end
+
+  defp do_create_token(owner_attr, application, opts) do
+    now = Keyword.get(opts, :now, DateTime.utc_now())
+
+    token_attrs =
+      %{
+        value: Base.hex_encode32(:crypto.strong_rand_bytes(32), padding: false),
+        application_id: application.id,
+        scopes: application.scopes,
+        expires_at: Timex.add(now, Timex.Duration.from_minutes(30)),
+        type: :access
+      }
+      |> Map.merge(owner_attr)
+
+    refresh_attrs =
+      if Keyword.get(opts, :create_refresh, true) do
+        %{
+          value: Base.hex_encode32(:crypto.strong_rand_bytes(32), padding: false),
+          application_id: application.id,
+          scopes: application.scopes,
+          # there's no real recourse when the refresh token expires and it's
+          # quite annoying, so make it "never" expire.
+          expires_at: Timex.add(now, Timex.Duration.from_days(365 * 100)),
+          type: :refresh,
+          refresh_token: nil
+        }
+        |> Map.merge(owner_attr)
+      else
+        nil
+      end
+
+    token_attrs = Map.put(token_attrs, :refresh_token, refresh_attrs)
+
+    %Token{}
+    |> Token.changeset(token_attrs)
+    |> Repo.insert()
+  end
+
+  # TODO: get_valid_token is basically the same as get_valid_code, refactor later
```

Comment:
I've amended the commit 3f44320e372aa79268a127968c04bbae24f6d4fd
I haven't merged the two functions since they are fairly small and that would be more confusing than anything. The main logic is already extracted into `check_expiry` and the query so I'm happy with how that looks now.